src/rust/ide/src/controller/project.rs

```diff
@@ -10,44 +10,145 @@ use crate::controller::Visualization;
+        let binary_protocol_events  = language_server_binary.event_stream();
+        let embedded_visualizations = default();
         let language_server_rpc     = Rc::new(language_server_client);
         let language_server_bin     = Rc::new(language_server_binary);
-        let logger                  = parent.sub("Project Controller");
-        let embedded_visualizations = default();
         let language_server         = language_server_rpc.clone();
         let visualization           = Visualization::new(language_server,embedded_visualizations);
-        Handle {module_registry,parser,language_server_rpc,language_server_bin,logger,visualization}
+
+        let ret = Handle {
+            project_name        : Rc::new(project_name.into()),
+            module_registry     : default(),
+            execution_contexts  : default(),
+            parser              : Parser::new_or_panic(),
+            language_server_rpc,
+            language_server_bin,
+            logger,
+            visualization,
+        };
+
+        let binary_handler = ret.binary_event_handler();
+        crate::executor::global::spawn(binary_protocol_events.for_each(binary_handler));
+        ret
+    }
+
+    /// Returns the primary content root id for this project.
+    pub fn content_root_id(&self) -> Uuid {
+        self.language_server_rpc.content_root()
+    }
+
+    /// Returns a handling function capable of processing updates from the binary protocol.
+    /// Such function will be then typically used to process events stream from the binary
+    /// connection handler.
+    pub fn binary_event_handler
+        (&self) -> impl Fn(enso_protocol::binary::Event) -> futures::future::Ready<()> {
+        let logger                  = self.logger.clone_ref();
+        let weak_execution_contexts = Rc::downgrade(&self.execution_contexts);
+        move |event| {
+            debug!(logger, "Received an event from the binary protocol: {event:?}");
+            use enso_protocol::binary::client::Event;
+            use enso_protocol::binary::Notification;
+            match event {
+                Event::Notification(Notification::VisualizationUpdate {context,data}) => {
+                    let data = VisualizationUpdateData::new(data);
+                    if let Some(execution_contexts) = weak_execution_contexts.upgrade() {
+                        let result = execution_contexts.dispatch_visualization_update(context,data);
+                        if let Err(error) = result {
+                            error!(logger,"Failed to handle the visualization update: {error}.");
+                        }
+                    } else {
+                        error!(logger,"Received a visualization update despite project being \
+                        already dropped.");
+                    }
+                }
+                _ => {} // At this point we do not care about anything other than visualizations.
```

Comment:
What about log/warning "unrecognized event"?

Reply:
Not really unrecognized, as the set of events is closed.
I've added explicit handling of all other cases and logging appropriate messages.